File: app/services/price_history_service.py
from datetime import date, timedelta
from typing import List, Optional
import logging

from sqlalchemy.orm import Session
from sqlalchemy.dialects.sqlite import insert

from app.models import Asset, PriceHistory
from app.services.quotes import (
    fetch_historical_quotes_batch,
    fetch_td_historical_quotes_batch,
)

logger = logging.getLogger(__name__)


def get_last_stored_date(asset_ticker: str, db: Session) -> Optional[date]:
    """Return the most recent date stored in price_history for this asset, or None."""
    row = (
        db.query(PriceHistory.date)
        .filter(PriceHistory.asset_id == asset_ticker)
        .order_by(PriceHistory.date.desc())
        .first()
    )
    return row[0] if row else None


def get_latest_price(asset_ticker: str, db: Session) -> Optional[float]:
    """Return the most recent close price stored for this asset, or None."""
    row = (
        db.query(PriceHistory.close)
        .filter(PriceHistory.asset_id == asset_ticker)
        .order_by(PriceHistory.date.desc())
        .first()
    )
    return row[0] if row else None


def upsert_prices(records: List[dict], db: Session) -> None:
    """
    Insert-or-update price_history rows.

    Each record must have keys: asset_id, date, close.
    Uses SQLite INSERT OR REPLACE via on_conflict_do_update.
    """
    if not records:
        return

    stmt = insert(PriceHistory).values(records)
    stmt = stmt.on_conflict_do_update(
        index_elements=["asset_id", "date"],
        set_={"close": stmt.excluded.close},
    )
    db.execute(stmt)
    db.commit()


def _first_transaction_date(asset: Asset) -> Optional[date]:
    """Return the earliest transaction date for this asset."""
    if not asset.transactions:
        return None
    return min(t.date for t in asset.transactions)


def backfill_asset(asset: Asset, db: Session) -> None:
    """
    Fetch and store close prices for a single asset from its first transaction
    date (or the day after the last stored date) up to today.
    """
    first_tx = _first_transaction_date(asset)
    if first_tx is None:
        return

    last_stored = get_last_stored_date(asset.ticker, db)
    start_date = (last_stored + timedelta(days=1)) if last_stored else first_tx
    today = date.today()

    if start_date > today:
        return

    try:
        if asset.type == "TESOURO_DIRETO":
            if not asset.yf_ticker:
                return
            raw = fetch_td_historical_quotes_batch([asset.yf_ticker])
            prices = raw.get(asset.yf_ticker, [])
        else:
            yf_ticker = asset.yf_ticker or (
                f"{asset.ticker}.SA" if "." not in asset.ticker else asset.ticker
            )
            raw = fetch_historical_quotes_batch({yf_ticker: asset.ticker}, start_date)
            prices = raw.get(asset.ticker, [])

        records = [
            {"asset_id": asset.ticker, "date": d, "close": close}
            for d, close in prices
            if d >= start_date
        ]
        upsert_prices(records, db)
        logger.info(f"Backfilled {len(records)} records for {asset.ticker}")
    except Exception as e:
        logger.error(f"Error backfilling {asset.ticker}: {e}")


def run_backfill(db_factory) -> None:
    """
    Backfill all assets with transactions, batching yfinance and TD assets separately
    to minimise API calls.
    """
    db: Session = db_factory()
    try:
        assets = db.query(Asset).all()
        # Pre-load transactions to avoid lazy-load issues outside the session
        for a in assets:
            _ = a.transactions  # noqa: F841

        yf_ticker_map: dict = {}  # {yf_ticker: asset_ticker}
        td_ticker_map: dict = {}  # {yf_ticker: asset}
        start_dates: dict = {}   # {asset_ticker: start_date}
        today = date.today()

        for asset in assets:
            first_tx = _first_transaction_date(asset)
            if first_tx is None:
                continue

            last_stored = get_last_stored_date(asset.ticker, db)
            start = (last_stored + timedelta(days=1)) if last_stored else first_tx
            if start > today:
                continue

            start_dates[asset.ticker] = start

            if asset.type == "TESOURO_DIRETO":
                if asset.yf_ticker:
                    td_ticker_map[asset.yf_ticker] = asset
            else:
                yf_ticker = asset.yf_ticker or (
                    f"{asset.ticker}.SA" if "." not in asset.ticker else asset.ticker
                )
                yf_ticker_map[yf_ticker] = asset.ticker

        # --- yfinance batch ---
        if yf_ticker_map:
            earliest = min(start_dates[t] for t in yf_ticker_map.values() if t in start_dates)
            try:
                batch = fetch_historical_quotes_batch(yf_ticker_map, earliest)
                records = []
                for asset_ticker, prices in batch.items():
                    cutoff = start_dates.get(asset_ticker, earliest)
                    for d, close in prices:
                        if d >= cutoff:
                            records.append({"asset_id": asset_ticker, "date": d, "close": close})
                upsert_prices(records, db)
                logger.info(f"Backfilled {len(records)} yfinance records for {len(yf_ticker_map)} assets")
            except Exception as e:
                logger.error(f"Error in yfinance backfill batch: {e}")

        # --- Tesouro Direto batch ---
        if td_ticker_map:
            try:
                batch = fetch_td_historical_quotes_batch(list(td_ticker_map.keys()))
                records = []
                for yf_ticker, prices in batch.items():
                    asset = td_ticker_map[yf_ticker]
                    cutoff = start_dates.get(asset.ticker, date.min)
                    for d, close in prices:
                        if d >= cutoff:
                            records.append({"asset_id": asset.ticker, "date": d, "close": close})
                upsert_prices(records, db)
                logger.info(f"Backfilled {len(records)} TD records for {len(td_ticker_map)} assets")
            except Exception as e:
                logger.error(f"Error in TD backfill batch: {e}")

    finally:
        db.close()


def run_daily_update(db_factory) -> None:
    """
    Fetch and store today's close price for all active assets.
    Uses the same batching strategy as run_backfill but only requests today's data.
    """
    db: Session = db_factory()
    try:
        assets = db.query(Asset).all()
        for a in assets:
            _ = a.transactions  # noqa: F841

        today = date.today()

        yf_ticker_map: dict = {}
        td_ticker_map: dict = {}

        for asset in assets:
            if not asset.transactions:
                continue
            if asset.type == "TESOURO_DIRETO":
                if asset.yf_ticker:
                    td_ticker_map[asset.yf_ticker] = asset
            else:
                yf_ticker = asset.yf_ticker or (
                    f"{asset.ticker}.SA" if "." not in asset.ticker else asset.ticker
                )
                yf_ticker_map[yf_ticker] = asset.ticker

        if yf_ticker_map:
            try:
                batch = fetch_historical_quotes_batch(yf_ticker_map, today)
                records = [
                    {"asset_id": asset_ticker, "date": d, "close": close}
                    for asset_ticker, prices in batch.items()
                    for d, close in prices
                    if d == today
                ]
                upsert_prices(records, db)
                logger.info(f"Daily update: stored {len(records)} yfinance prices")
            except Exception as e:
                logger.error(f"Error in daily yfinance update: {e}")

        if td_ticker_map:
            try:
                batch = fetch_td_historical_quotes_batch(list(td_ticker_map.keys()))
                records = []
                for yf_ticker, prices in batch.items():
                    asset = td_ticker_map[yf_ticker]
                    for d, close in prices:
                        if d == today:
                            records.append({"asset_id": asset.ticker, "date": d, "close": close})
                upsert_prices(records, db)
                logger.info(f"Daily update: stored {len(records)} TD prices")
            except Exception as e:
                logger.error(f"Error in daily TD update: {e}")
    finally:
        db.close()

File: src/main/kotlin/com/stocks/service/PriceHistoryService.kt
package com.stocks.service

import com.stocks.dto.NO_QUOTE_TYPES
import com.stocks.model.*
import org.jetbrains.exposed.v1.core.*
import org.jetbrains.exposed.v1.jdbc.*
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate

// Resolves Yahoo Finance ticker from asset ticker
fun resolveYfTicker(
    ticker: String,
    yfTicker: String?
): String = yfTicker ?: classifyTicker(ticker).yfCandidates.first()

// Data class for asset categorization input
data class AssetTickerInfo(
    val ticker: String,
    val yfTicker: String?,
    val type: String?,
    val delisted: Boolean = false,
)

// Result of categorization
data class TickerMaps(
    val yfTickerMap: Map<String, String>, // yfTicker -> assetTicker
    val tdTickerMap: Map<String, String>, // yfTicker -> assetTicker
)

// Categorizes assets into Yahoo Finance vs Tesouro Direto maps
fun categorizeAssets(assets: List<AssetTickerInfo>): TickerMaps {
    val yfTickerMap = mutableMapOf<String, String>()
    val tdTickerMap = mutableMapOf<String, String>()

    for (asset in assets) {
        if (asset.delisted) continue
        if (asset.type in NO_QUOTE_TYPES) continue
        if (asset.type == "TESOURO_DIRETO") {
            if (asset.yfTicker != null) {
                tdTickerMap[asset.yfTicker] = asset.ticker
            }
        } else {
            val resolved = resolveYfTicker(asset.ticker, asset.yfTicker)
            yfTickerMap[resolved] = asset.ticker
        }
    }

    return TickerMaps(yfTickerMap, tdTickerMap)
}

// Filters batch API results into PriceRecords
fun filterBatchToRecords(
    batch: Map<String, List<Pair<LocalDate, Double>>>,
    tickerResolver: (String) -> String? = { it },
    datePredicate: (String, LocalDate) -> Boolean,
): List<PriceRecord> {
    val records = mutableListOf<PriceRecord>()
    for ((key, prices) in batch) {
        val assetTicker = tickerResolver(key) ?: continue
        for ((date, close) in prices) {
            if (datePredicate(assetTicker, date)) {
                records.add(PriceRecord(assetTicker, date, close))
            }
        }
    }
    return records
}

@Service
class PriceHistoryService(
    private val quoteService: QuoteService,
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    fun getLatestPrice(assetTicker: String): Double? =
        transaction {
            PriceHistories
                .select(PriceHistories.close)
                .where { PriceHistories.assetId eq assetTicker }
                .orderBy(PriceHistories.date, SortOrder.DESC)
                .limit(1)
                .firstOrNull()
                ?.get(PriceHistories.close)
        }

    fun getPricesForDate(
        assetTickers: List<String>,
        date: LocalDate
    ): Map<String, Double> =
        transaction {
            PriceHistories
                .select(PriceHistories.assetId, PriceHistories.close)
                .where {
                    (PriceHistories.assetId inList assetTickers) and
                        (PriceHistories.date eq date)
                }.associate { it[PriceHistories.assetId] to it[PriceHistories.close] }
        }

    fun getLastStoredDate(assetTicker: String): LocalDate? =
        transaction {
            PriceHistories
                .select(PriceHistories.date)
                .where { PriceHistories.assetId eq assetTicker }
                .orderBy(PriceHistories.date, SortOrder.DESC)
                .limit(1)
                .firstOrNull()
                ?.get(PriceHistories.date)
        }

    fun upsertPrices(records: List<PriceRecord>) {
        if (records.isEmpty()) return
        transaction {
            for (record in records) {
                val existing =
                    PriceHistories
                        .selectAll()
                        .where {
                            (PriceHistories.assetId eq record.assetId) and
                                (PriceHistories.date eq record.date)
                        }.firstOrNull()

                if (existing != null) {
                    PriceHistories.update({
                        (PriceHistories.assetId eq record.assetId) and
                            (PriceHistories.date eq record.date)
                    }) {
                        it[close] = record.close
                    }
                } else {
                    PriceHistories.insert {
                        it[assetId] = record.assetId
                        it[date] = record.date
                        it[close] = record.close
                    }
                }
            }
        }
    }

    fun generateDelistedPrices(assetTicker: String) {
        val txPrices =
            transaction {
                TransactionEntity
                    .find { Transactions.assetId eq assetTicker }
                    .toList()
                    .sortedBy { it.date }
                    .map { it.date to it.price }
            }
        if (txPrices.isEmpty()) return

        val today = LocalDate.now()
        val records = mutableListOf<PriceRecord>()

        if (txPrices.size == 1) {
            val (startDate, price) = txPrices.first()
            var d = startDate
            while (!d.isAfter(today)) {
                records.add(PriceRecord(assetTicker, d, price))
                d = d.plusDays(1)
            }
        } else {
            for (i in 0 until txPrices.size - 1) {
                val (dateA, priceA) = txPrices[i]
                val (dateB, priceB) = txPrices[i + 1]
                val totalDays =
                    java.time.temporal.ChronoUnit.DAYS
                        .between(dateA, dateB)
                var d = dateA
                while (d.isBefore(dateB)) {
                    val dayOffset =
                        java.time.temporal.ChronoUnit.DAYS
                            .between(dateA, d)
                    val price =
                        if (totalDays > 0) {
                            priceA + (priceB - priceA) * dayOffset.toDouble() / totalDays.toDouble()
                        } else {
                            priceA
                        }
                    records.add(PriceRecord(assetTicker, d, price))
                    d = d.plusDays(1)
                }
            }
            val lastPrice = txPrices.last().second
            var d = txPrices.last().first
            while (!d.isAfter(today)) {
                records.add(PriceRecord(assetTicker, d, lastPrice))
                d = d.plusDays(1)
            }
        }

        upsertPrices(records)
        logger.debug("Generated ${records.size} delisted prices for $assetTicker")
    }

    fun runBackfill() {
        transaction {
            val assets = AssetEntity.all().toList()
            val today = LocalDate.now()
            val startDates = mutableMapOf<String, LocalDate>()

            // Process delisted assets separately
            val delistedAssets = assets.filter { it.delisted && !it.transactions.empty() }
            for (asset in delistedAssets) {
                generateDelistedPrices(asset.ticker.value)
            }

            val assetInfoList =
                assets.mapNotNull { asset ->
                    val ticker = asset.ticker.value
                    val firstTx = asset.transactions.toList().minOfOrNull { it.date } ?: return@mapNotNull null
                    val lastStored = getLastStoredDate(ticker)
                    val fiveYearsAgo = today.minusYears(5)
                    val start = if (lastStored != null) lastStored.plusDays(1) else minOf(firstTx, fiveYearsAgo)
                    if (start > today) return@mapNotNull null
                    startDates[ticker] = start
                    AssetTickerInfo(ticker, asset.yfTicker, asset.type, asset.delisted)
                }

            val maps = categorizeAssets(assetInfoList)

            // yfinance batch
            if (maps.yfTickerMap.isNotEmpty()) {
                try {
                    val earliest =
                        maps.yfTickerMap.values
                            .mapNotNull { startDates[it] }
                            .minOrNull() ?: today
                    val batch = quoteService.fetchHistoricalQuotesBatch(maps.yfTickerMap, earliest)
                    val records =
                        filterBatchToRecords(batch) { ticker, date ->
                            date >= (startDates[ticker] ?: earliest)
                        }
                    upsertPrices(records)
                    logger.info("Backfilled ${records.size} yfinance records for ${maps.yfTickerMap.size} assets")
                } catch (e: Exception) {
                    logger.error("Error in yfinance backfill batch: ${e.message}")
                }
            }

            // Tesouro Direto batch
            if (maps.tdTickerMap.isNotEmpty()) {
                try {
                    val batch = quoteService.fetchTdHistoricalQuotesBatch(maps.tdTickerMap.keys.toList())
                    val records =
                        filterBatchToRecords(batch, { maps.tdTickerMap[it] }) { ticker, date ->
                            date >= (startDates[ticker] ?: LocalDate.MIN)
                        }
                    upsertPrices(records)
                    logger.info("Backfilled ${records.size} TD records for ${maps.tdTickerMap.size} assets")
                } catch (e: Exception) {
                    logger.error("Error in TD backfill batch: ${e.message}")
                }
            }
        }
    }

    fun runDailyUpdate() {
        transaction {
            val assets = AssetEntity.all().toList()
            val today = LocalDate.now()

            // Process delisted assets separately
            val delistedAssets = assets.filter { it.delisted && !it.transactions.empty() }
            for (asset in delistedAssets) {
                generateDelistedPrices(asset.ticker.value)
            }

            val assetInfoList =
                assets
                    .filter { !it.transactions.empty() }
                    .map { AssetTickerInfo(it.ticker.value, it.yfTicker, it.type, it.delisted) }

            val maps = categorizeAssets(assetInfoList)

            if (maps.yfTickerMap.isNotEmpty()) {
                try {
                    val batch = quoteService.fetchHistoricalQuotesBatch(maps.yfTickerMap, today)
                    val records =
                        filterBatchToRecords(batch) { _, date -> date == today }
                    upsertPrices(records)
                    logger.info("Daily update: stored ${records.size} yfinance prices")
                } catch (e: Exception) {
                    logger.error("Error in daily yfinance update: ${e.message}")
                }
            }

            if (maps.tdTickerMap.isNotEmpty()) {
                try {
                    val batch = quoteService.fetchTdHistoricalQuotesBatch(maps.tdTickerMap.keys.toList())
                    val records =
                        filterBatchToRecords(batch, { maps.tdTickerMap[it] }) { _, date -> date == today }
                    upsertPrices(records)
                    logger.info("Daily update: stored ${records.size} TD prices")
                } catch (e: Exception) {
                    logger.error("Error in daily TD update: ${e.message}")
                }
            }
        }
    }
}

data class PriceRecord(
    val assetId: String,
    val date: LocalDate,
    val close: Double,
)
